### PR TITLE
SampleGen: generate skeletons for C# standalone samples

### PR DESCRIPTION
**Pull request by hzyi-google**
Add a `standalone_sample.snip` for CSharp to start working on generating standalone samples, similar to [this one](https://github.com/SurferJeffAtGoogle/csharp-docs-samples/blob/aa405e94f5e19da8d108336bffe5ddb2ac0ee0b6/dlp/api/Snippets/DeidentifyMasking.cs).

### CODE REVIEW

**Comment by hzyi-google** [comment]
@googleapis/samplegen @chrisdunelm PTAL

File: src/main/java/com/google/api/codegen/gapic/GapicGeneratorFactory.java [hunk 1]
```diff
@@ -184,6 +185,9 @@
               newCsharpGenerator.apply(
                   CSharpBasicPackageTransformer.forUnitTests(unitTestPathMapper)));
         }
+        if (devSamples) {
+          generators.add(newCsharpGenerator.apply(new CSharpStandaloneSampleTransformer()));
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/gapic/GapicGeneratorFactory.java]: Can we call this `CSharpGapicSamplesTransformer` for consistency with the other languages, or does this conflict  with an existing class?

File: src/main/java/com/google/api/codegen/transformer/StaticLangApiMethodTransformer.java [hunk 2]
```diff
@@ -68,6 +69,14 @@ public StaticLangApiMethodTransformer() {
     this(SampleTransformer.create(SampleType.IN_CODE));
   }
 
+  // TODO: Currently overrode in CSharpApiMethodTransformer. Inspect whether the same logic applies
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/transformer/StaticLangApiMethodTransformer.java]: nit: "overridden"

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpApiMethodTransformer.java [hunk 3]
```diff
@@ -86,4 +110,118 @@ protected void setServiceResponseTypeName(
...
+            .addAll(csharpCommonTransformer.pagedMethodAdditionalParams())
+            .addAll(csharpCommonTransformer.callSettingsParam())
+            .build();
+
+    List<StaticLangApiMethodView> apiMethods = new ArrayList<>();
+    for (MethodModel method : csharpCommonTransformer.getSupportedMethods(context)) {
+      MethodConfig methodConfig = context.getMethodConfig(method);
+      MethodContext requestMethodContext = context.asRequestMethodContext(method);
+      if (methodConfig.isGrpcStreaming()) {
+        // Only for protobuf-based APIs.
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/transformer/csharp/CSharpApiMethodTransformer.java]: Why do we need this comment? (I realize this is moved from the `CSharpGapicClientTransformer`, but I'm still curious)

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpApiMethodTransformer.java [hunk 4]
```diff
@@ -86,4 +110,118 @@ protected void setServiceResponseTypeName(
...
+            GapicMethodContext methodContext =
+                context.asFlattenedMethodContext(requestMethodContext, flatteningGroup);
+            apiMethods.add(
+                generateGrpcStreamingFlattenedMethod(
+                    methodContext, csharpCommonTransformer.callSettingsParam()));
+          }
+        }
+        apiMethods.add(generateGrpcStreamingRequestObjectMethod(requestMethodContext));
+      } else if (requestMethodContext.isLongRunningMethodContext()) {
+        // Only for protobuf-based APIs.
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/transformer/csharp/CSharpApiMethodTransformer.java]: Why this comment?

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpStandaloneSampleTransformer.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+import com.google.common.collect.Streams;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.List;
+
+/** A transformer that generates C# standalone samples. */
+public class CSharpStandaloneSampleTransformer implements ModelToViewTransformer<ProtoApiModel> {
+
+  private static final String STANDALONE_SAMPLE_TEMPLATE_FILENAME = "csharp/standalone_sample.snip";
+  private static final String CSHARP_SAMPLE_PACKAGE_NAME = "Snippets";
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/transformer/csharp/CSharpStandaloneSampleTransformer.java]: Why `"Snippets"`? Maybe it's more consistent to call it "Samples" or "StandaloneSamples". I think (not sure, @chrisdunelm can confirm) that "snippets" is the term C# is using for their current version of in-code samples.

**Comment by chrisdunelm** [comment]
Apologies for not reviewing yet, I've been busier than I expected. I'll do it tomorrow if you can wait another day.

**Comment by hzyi-google** [comment]
@chrisdunelm No worries at all.

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpApiMethodTransformer.java (hunk 3 above)
Comment (hzyi-google) [reply src/main/java/com/google/api/codegen/transformer/csharp/CSharpApiMethodTransformer.java]: I  _think_ that means it does not apply to discogapics.
That said, looking a few lines above, we have already checked `interfaceContext` is a `GapicInterfaceContext`. So I do think this comment is redundant. I'll remove it in a coming commit.
(Also noticed the error message of that `checkArgument` is not appropriate, will change that too)

File: src/main/java/com/google/api/codegen/gapic/GapicGeneratorFactory.java (hunk 1 above)
Comment (hzyi-google) [reply src/main/java/com/google/api/codegen/gapic/GapicGeneratorFactory.java]: Yeah, CSharp has a `CSharpGapicSnippetsTransformer`. It doesn't conflict with `CSharpGapicSamplesTransformer`, but does cause some confusion (Snippets? Samples?) IMO. So I decided to rename the transformer for standalone samples.

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpStandaloneSampleTransformer.java (hunk 5 above)
Comment (hzyi-google) [reply src/main/java/com/google/api/codegen/transformer/csharp/CSharpStandaloneSampleTransformer.java]: The name is from @beccasaurus 's [canonical sample](https://github.com/SurferJeffAtGoogle/csharp-docs-samples/blob/aa405e94f5e19da8d108336bffe5ddb2ac0ee0b6/dlp/api/Snippets/DeidentifyMasking.cs). But I do see the confusion with incode samples here :/

File: src/main/resources/com/google/api/codegen/csharp/standalone_sample.snip [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+@extends "csharp/common_initcode.snip"
+
+# StaticLangFileView
+@snippet generate(sampleFile)
+  @let sampleClass = sampleFile.classView
+    @let apiMethod = sampleClass.libraryMethod, sample = sampleClass.sample
+  	{@initialComments(sampleFile.fileHeader)}
+    // DO NOT EDIT! This is a generated sample ("{@sample.callingForm}",  "{@sample.valueSet.id}")
+
+    // [START {@sample.regionTag}]
```
Comment (chrisdunelm) [inline src/main/resources/com/google/api/codegen/csharp/standalone_sample.snip]: nit: `[END ...]` missing

File: src/main/resources/com/google/api/codegen/csharp/standalone_sample.snip [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    // FIXME: import section
+
+    class {@sampleClass.name}
+    {
+      // FIXME: sample function documentation
+      static public void {@sample.sampleFunctionName}({@commaSeparatedArgAndTypes(sample.sampleInitCode.argDefaultParams)})
```
Comment (chrisdunelm) [inline src/main/resources/com/google/api/codegen/csharp/standalone_sample.snip]: Please change the modifier order: `public static`.
(and below)

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpStandaloneSampleTransformer.java (hunk 5 above)
Comment (vchudnov-g) [reply src/main/java/com/google/api/codegen/transformer/csharp/CSharpStandaloneSampleTransformer.java]: Let's not block this PR, but let's double check whether we  can use "samples" for consistency

Comment (beccasaurus) [reply src/main/java/com/google/api/codegen/transformer/csharp/CSharpStandaloneSampleTransformer.java]: Package name... hmm. Not passionate at all, developers don't see it.

How about: `Language.v1beta2.Samples.csproj`?

`Talent.v4beta1.Samples.csproj`
`Speech.v1.Samples.csproj`

Comment (hzyi-google) [reply src/main/java/com/google/api/codegen/transformer/csharp/CSharpStandaloneSampleTransformer.java]: @beccasaurus I'm gonna change it to `Samples` for now. Let's reconsider the structure when working on canonical samples?

**Comment by hzyi-google** [comment]
Thanks for reviewing!

**Pull request merged**